samples: Bluetooth: hci_ipc: Support two Broadcast ISO instances

cvinayak opened the pull request:
Update to support two Broadcast ISO instances and update HCI
buffer counts.

PTS test variants use so, hence be in sync for samples too.

Fix assertion on BIG terminate sync and due to not receiving
terminate control procedure when there are no free buffers.
Discovered during this PR.

cvinayak committed:
tests: bsim: Bluetooth: Workaround Zephyr Controller scheduling overlap

Workaround test failures due to Zephyr Controller scheduling
overlap causing Periodic Advertising Sync timeouts, ACL
connection failed to be established or Controller assertion
when having latency at the CIS setup instant.

Workaround by using `-start_offset` to ensure that the ACL
established does not overlap the sink being established.

Relates to commit 490da9ec9613 ("tests: bsim: Bluetooth:
Workaround overlapping ACL in BAP broadcast sink").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: hci_ipc: Support two Broadcast ISO instances

Update to support two Broadcast ISO instances and update HCI
buffer counts.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix assertion on BIG terminate sync

Fix assertion on BIG terminate sync and due to not receiving
terminate control procedure when there are no free buffers.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Missing Kconfig depends for BT_BUF_EVT_DISCARDABLE_

Missing depends on !BT_HCI_RAW for HCI Discardable Events
count and size.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Move BT_BUF_EVT_DISCARDABLE_ to Host Kconfig

Move BT_BUF_EVT_DISCARDABLE_ to Host Kconfig, as this is
not used by the Controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented:
cc @alexsven this may fix the issue you saw for the BAP test with 2 BIGs

cvinayak commented:
I have 2 challenges in this PR... the RAM overflows by ~ 360 bytes... and I could reduce RAM by reducing to `CONFIG_BT_BUF_ACL_TX_COUNT=3`, but then babblesim test fails with Controller assertion!... needs investigation

Thalley reviewed (commented):
Overall changes look good to me, but the memory issue is tricky :)

Thalley commented:
Fyi, this is no longer really required by anyone after https://github.com/zephyrproject-rtos/zephyr/pull/88300 was merged as that supports it

cvinayak commented:
> Fyi, this is no longer really required by anyone after #88300 was merged as that supports it

The PR started as an analyse of many audio tests failing when buffer counts where changed. Example, `CONFIG_BT_BUF_ACL_TX_COUNT=3`  causes failures as seen in this PR now.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -719,6 +712,15 @@ static void isr_rx(void *param)
+		if (ull_iso_pdu_rx_alloc_peek(2U) == NULL) {
+			goto isr_rx_done;
+		}
+
```
1 buffer check to ensure we pick up the new control sequence number, 2 buffers is checked to decide if SDU to be buffered.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -699,18 +699,11 @@ static void isr_rx(void *param)
 		}
 
+		node_rx = ull_iso_pdu_rx_alloc_peek(1U);
+		LL_ASSERT(node_rx);
```
There should always be a  buffer setup in the radio h/w from this pool. This is a fatal assertion if there is none.

cvinayak commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -166,10 +167,10 @@ config BT_BUF_EVT_DISCARDABLE_SIZE
+	depends on !BT_HCI_RAW
 	range 1 $(UINT8_MAX)
 	default 20 if BT_MESH
 	default 3
```
done

PavelVPV commented on tests/bsim/bluetooth/ll/multiple_id/prj.conf:
```diff
@@ -12,19 +12,20 @@ CONFIG_BT_AUTO_DATA_LEN_UPDATE=y
 # using ticker nodes count up to 255 (UINT8_MAX) in the hci_ipc with ISO support
-CONFIG_BT_MAX_CONN=231
-CONFIG_BT_ID_MAX=231
+CONFIG_BT_MAX_CONN=230
```
Should the multiple_id test change be in a separate commit? It is also not clear why it is changed.

cvinayak replied:
No, keeping it in the same commit as `CONFIG_BT_CTLR_ADV_ISO_SET=2` is an increase by one, hence reducing the number of supported simultaneous connections by one.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -699,18 +699,11 @@ static void isr_rx(void *param)
 		}
 
+		node_rx = ull_iso_pdu_rx_alloc_peek(1U);
+		LL_ASSERT(node_rx);
```
This is fatal assert, as a node_rx shall be present as it was provided to the radio ptr before the ISR executes.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -719,6 +712,15 @@ static void isr_rx(void *param)
+		 * is to ensure a PDU can be setup for the radio DMA to
+		 * receive in the next sub_interval/iso_interval.
+		 */
+		if (ull_iso_pdu_rx_alloc_peek(2U) == NULL) {
```
Outside the scope, as will be good to do the request in separate PR.

cvinayak committed:
tests: bsim: Bluetooth: Test 250 ACL connections

Updated to test 250 ACL connections in the multiple
identity across nrf52_bsim, nrf5340bsim and nrf54l15bsim
targets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

jhedberg commented on subsys/bluetooth/common/Kconfig:
```diff
@@ -150,6 +150,7 @@ config BT_BUF_EVT_RX_COUNT
 
 config BT_BUF_EVT_DISCARDABLE_SIZE
 	int "Maximum supported discardable HCI Event buffer length"
+	depends on !BT_HCI_RAW
```
What's the reason these are in `common/` instead of `host/`? If they're host specific they should probably be moved, and then you don't need the explicit exclusion of HCI_RAW.

cvinayak replied:
Agree. Sure will give a try, and get back.
